src/index.ts

```diff
@@ -0,0 +1,522 @@
+    ) {
+      if (isModuleId(moduleIdOrDependencyPathList)) {
+        const module = moduleMap.get(moduleIdOrDependencyPathList);
+        if (!module || module.state !== ModuleState.INITIALIZED) {
+          throw Error("#3 " + moduleIdOrDependencyPathList);
+        }
+        return module.exports;
+      }
+      const dependencyIds = moduleIdOrDependencyPathList.map(modulePathToId);
+      let remainingDependencies = moduleIdOrDependencyPathList.length + 1;
+      const rId = requestId || ((Math.random() + "") as RequestId);
+
+      function ensureCommonJsDependencies() {
+        let cjsModule: { exports: unknown } = { exports: {} };
+        const module = moduleMap.get(moduleId);
+        if (module) {
+          if (module.state != ModuleState.WAITING_FOR_DEPENDENCIES) {
+            return panic("Unexpected module state");
+          }
+          cjsModule = module;
+        }
+        moduleMap.set("require", {
+          id: "require",
+          state: ModuleState.INITIALIZED,
+          exports: require
+        });
+        moduleMap.set("exports", {
+          id: "exports",
+          state: ModuleState.INITIALIZED,
+          exports: cjsModule.exports
+        });
+        moduleMap.set("module", {
+          id: "module",
+          state: ModuleState.INITIALIZED,
+          exports: cjsModule
+        });
+      }
+
+      function dependencyReadyCallback() {
+        if (--remainingDependencies == 0) {
+          ensureCommonJsDependencies();
+          try {
+            const dependencies: unknown[] = dependencyIds.map(function(id) {
+              const module = moduleMap.get(id);
+              if (!module) {
+                return panic(
+                  "Mismatch in reported and actually loaded modules"
+                );
+              }
+              if (
+                module.state == ModuleState.NETWORK_LOADING ||
+                module.state == ModuleState.DEFINED
+              ) {
+                return panic(
+                  "Unexpected module state when resolving dependencies"
+                );
+              }
+              if (module.state == ModuleState.ERROR) {
+                throw Error("#4 " + id);
+              }
+              return module.exports;
+            });
+            (onSuccess || noop).apply(undefined, dependencies);
+          } catch (error) {
+            (onError || noop)(error);
+          }
+        }
+      }
+
+      ensureCommonJsDependencies();
+      dependencyIds.forEach(function(id) {
+        request(id, rId, dependencyReadyCallback);
+      });
+      dependencyReadyCallback();
+    } as IamdeeRequireFunction;
+    require["config"] = config;
+    return require;
+  }
+
+  function loadModule(
+    id: ModuleId,
+    src: SourceUrl,
+    requestId: RequestId,
+    callback: ModuleCallback
+  ): NetworkLoadingModule {
+    // need to add new script to the browser
+    const el = doc.createElement("script");
+    elementMap.set(el, id);
+    el.src = src;
```

Comment:
even-though `async` is by default for dynamic scripts, its good to state that in comments here.

Reply:
good call, added